File: app/src/main/java/de/westnordost/streetcomplete/view/ItemViewHolder.java
package de.westnordost.streetcomplete.view;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import de.westnordost.streetcomplete.R;

public class ItemViewHolder  extends RecyclerView.ViewHolder
{
	private ImageView imageView;
	@Nullable private TextView textView;
	@Nullable private TextView descriptionView;

	public interface OnClickListener
	{
		void onClick(int index);
	}

	public ItemViewHolder(@NonNull View itemView)
	{
		super(itemView);
		imageView = itemView.findViewById(R.id.imageView);
		textView = itemView.findViewById(R.id.textView);
		descriptionView = itemView.findViewById(R.id.descriptionView);
	}

	public void bind(@NonNull Item item)
	{
		imageView.setImageResource(item.drawableId);

		if(textView != null)
		{
			if (item.titleId != 0) textView.setText(item.titleId);
			else textView.setText(null);
		}

		if(descriptionView != null)
		{
			boolean hasDescription = item.descriptionId != 0;
			descriptionView.setVisibility(hasDescription ? View.VISIBLE : View.GONE);
			if (hasDescription) descriptionView.setText(item.descriptionId);
			else descriptionView.setText(null);
		}
	}

	public void setSelected(boolean selected)
	{
		itemView.setSelected(selected);
	}

	public void setOnClickListener(OnClickListener listener)
	{
		if(listener == null) itemView.setOnClickListener(null);
		itemView.setOnClickListener(v -> {
			int index = getAdapterPosition();
			if(index != RecyclerView.NO_POSITION) listener.onClick(index);
		});
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/view/ItemViewHolder.kt
package de.westnordost.streetcomplete.view

import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.core.view.isGone
import androidx.recyclerview.widget.RecyclerView
import de.westnordost.streetcomplete.R

class ItemViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

    private val imageView: ImageView = itemView.findViewById(R.id.imageView)
    private val textView: TextView? = itemView.findViewById(R.id.textView)
    private val descriptionView: TextView? = itemView.findViewById(R.id.descriptionView)

    var isSelected: Boolean
        get() = itemView.isSelected
        set(value) { itemView.isSelected = value }

    var onClickListener: ((index: Int) -> Unit)? = null
        set(value) {
            field = value
            if (value == null) itemView.setOnClickListener(null)
            else itemView.setOnClickListener {
                val index = adapterPosition
                if (index != RecyclerView.NO_POSITION) value.invoke(index)
            }
        }

    fun bind(item: Item<*>) {
        val drawableId = item.drawableId
        if (drawableId != null)
            imageView.setImageResource(drawableId)
        else
            imageView.setImageDrawable(null)

        val titleId = item.titleId
        if (titleId != null)
            textView?.setText(titleId)
        else
            textView?.text = null

        val descriptionId = item.descriptionId
        descriptionView?.isGone = descriptionId == null
        if (descriptionId != null)
            descriptionView?.setText(descriptionId)
        else
            descriptionView?.text = null
    }
}
